### ISSUE TITLE
Perl_cast_uv broken with CASTFLAGS=7, no quad math, uvsize=4

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15162]
Migrated from [rt.perl.org#127474](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=127474) (status was 'resolved')

Searchable as RT127474$

**Comment by p5pRT** [comment]
### From pipcet@gmail.com

#### Created by pipcet@gmail.com



TL;DR&#8203;: patch&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/numeric.c b/numeric.c
index f1de219..f645502 100644
--- a/numeric.c
+++ b/numeric.c
@@ -39,7 +39,7 @@ Perl_cast_ulong(NV f)
     if (f < U32_MAX_P1_HALF)
       return (U32) f;
     f -= U32_MAX_P1_HALF;
-    return ((U32) f) | (1 + U32_MAX >> 1);
+    return ((U32) f) | (1 + (U32_MAX >> 1));
 #else
     return (U32) f;
 #endif
@@ -57,7 +57,7 @@ Perl_cast_i32(NV f)
     if (f < U32_MAX_P1_HALF)
       return (I32)(U32) f;
     f -= U32_MAX_P1_HALF;
-    return (I32)(((U32) f) | (1 + U32_MAX >> 1));
+    return (I32)(((U32) f) | (1 + (U32_MAX >> 1)));
 #else
     return (I32)(U32) f;
 #endif
@@ -76,7 +76,7 @@ Perl_cast_iv(NV f)
     if (f < UV_MAX_P1_HALF)
       return (IV)(UV) f;
     f -= UV_MAX_P1_HALF;
-    return (IV)(((UV) f) | (1 + UV_MAX >> 1));
+    return (IV)(((UV) f) | (1 + (UV_MAX >> 1)));
 #else
     return (IV)(UV) f;
 #endif
@@ -94,7 +94,7 @@ Perl_cast_uv(NV f)
     if (f < UV_MAX_P1_HALF)
       return (UV) f;
     f -= UV_MAX_P1_HALF;
-    return ((UV) f) | (1 + UV_MAX >> 1);
+    return ((UV) f) | (1 + (UV_MAX >> 1));
 #else
     return (UV) f;
 #endif


```
</details>
I am trying to build Perl 5 (git blead) using a somewhat unusual
virtual machine for which config\.sh has decided in its infinite wisdom
castflags should be 7; I have decided quadmath should not be defined;
and the size of a UV is 4 bytes\. This combination is running into
problems\.

The symptom is that t/base/lex\.t fails with the error message

&nbsp;    not ok 7 :443541004&#8203;:

generated by this code

&nbsp;    eval '$foo = 123\+123\.4\+123e4\+123\.4E5\+123\.4e\+5\+\.12;';

&nbsp;    $foo = int\($foo \* 100 \+ \.5\);
&nbsp;    if \($foo eq 2591024652\) \{print "ok 7\\n";\} else \{print "not ok 7
:$foo&#8203;:\\n";\}

We're off by 1\<\<31\! The culprit is in numeric\.c's Perl\_cast\_uv\, which
contains this line \(enabled when CASTFLAGS & 2\, as is the case here\)&#8203;:

&nbsp;    return \(\(UV\) f\) | \(1 \+ UV\_MAX >> 1\);

I think that should be

&nbsp;    return \(\(UV\) f\) | \(1 \+ \(UV\_MAX >> 1\)\);

which is equivalent on systems for which UV\_MAX \+ 1 isn't 0 \(assuming
UV\_MAX is odd\, at least\.\.\.\)\, but works on systems for which UV\_MAX \+ 1
is 0\.  The original line also causes compiler warnings along the lines
of&#8203;:

&nbsp;    numeric\.c&#8203;:79&#8203;:40&#8203;: warning&#8203;: suggest parentheses around ‘\+’ inside ‘>>’
\[\-Wparentheses\]
&nbsp;         return \(IV\)\(\(\(UV\) f\) | \(1 \+ UV\_MAX >> 1\)\);

So even if I'm wrong\, this code should probably be fixed another way
in order to avoid the warning to keep compile\-time noise down\.

\(It's very suspicious that I'm using gcc but castflags has been set to
7\, but I think we should err on the side of allowing castflags to be
set that needn't be\)\.

Unfortunately\, perlbug doesn't run yet on the virtual machine\, so I'm
attaching config\.sh instead of the build configuration for my system
Perl\.

The GCC backend I'm using is experimental and not really ready for public
discussion yet\, but I'd be happy to share further details that are needed
to fix this bug\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low


```
</details>

**Comment by p5pRT** [comment]
### From pipcet@gmail.com

<details><summary>perl-002.diff</summary>

```diff
diff --git a/numeric.c b/numeric.c
index f1de219..f645502 100644
--- a/numeric.c
+++ b/numeric.c
@@ -39,7 +39,7 @@ Perl_cast_ulong(NV f)
     if (f < U32_MAX_P1_HALF)
       return (U32) f;
     f -= U32_MAX_P1_HALF;
-    return ((U32) f) | (1 + U32_MAX >> 1);
+    return ((U32) f) | (1 + (U32_MAX >> 1));
 #else
     return (U32) f;
 #endif
@@ -57,7 +57,7 @@ Perl_cast_i32(NV f)
     if (f < U32_MAX_P1_HALF)
       return (I32)(U32) f;
     f -= U32_MAX_P1_HALF;
-    return (I32)(((U32) f) | (1 + U32_MAX >> 1));
+    return (I32)(((U32) f) | (1 + (U32_MAX >> 1)));
 #else
     return (I32)(U32) f;
 #endif
@@ -76,7 +76,7 @@ Perl_cast_iv(NV f)
     if (f < UV_MAX_P1_HALF)
       return (IV)(UV) f;
     f -= UV_MAX_P1_HALF;
-    return (IV)(((UV) f) | (1 + UV_MAX >> 1));
+    return (IV)(((UV) f) | (1 + (UV_MAX >> 1)));
 #else
     return (IV)(UV) f;
 #endif
@@ -94,7 +94,7 @@ Perl_cast_uv(NV f)
     if (f < UV_MAX_P1_HALF)
       return (UV) f;
     f -= UV_MAX_P1_HALF;
-    return ((UV) f) | (1 + UV_MAX >> 1);
+    return ((UV) f) | (1 + (UV_MAX >> 1));
 #else
     return (UV) f;
 #endif

```
</details>

**Comment by p5pRT** [comment]
### From pipcet@gmail.com

[config.sh](https://rt-archive.perl.org/perl5/Ticket/Attachment/1387818/746611/config.sh)

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri Feb 05 21&#8203;:31&#8203;:07 2016\, pipcet@&#8203;gmail\.com wrote&#8203;:
> TL;DR&#8203;: patch&#8203;:
> 
> diff \-\-git a/numeric\.c b/numeric\.c
> index f1de219\.\.f645502 100644
> \-\-\- a/numeric\.c
> \+\+\+ b/numeric\.c
\.\.\.
> 
> I am trying to build Perl 5 \(git blead\) using a somewhat unusual
> virtual machine for which config\.sh has decided in its infinite wisdom
> castflags should be 7; I have decided quadmath should not be defined;
> and the size of a UV is 4 bytes\. This combination is running into
> problems\.
> 
> The symptom is that t/base/lex\.t fails with the error message
> 
>     not ok 7 :443541004&#8203;:
> 
> generated by this code
> 
>     eval '$foo = 123\+123\.4\+123e4\+123\.4E5\+123\.4e\+5\+\.12;';
> 
>     $foo = int\($foo \* 100 \+ \.5\);
>     if \($foo eq 2591024652\) \{print "ok 7\\n";\} else \{print "not ok 7
> :$foo&#8203;:\\n";\}
> 
> We're off by 1\<\<31\! The culprit is in numeric\.c's Perl\_cast\_uv\, which
> contains this line \(enabled when CASTFLAGS & 2\, as is the case here\)&#8203;:

I didn't get the exact failures you did by manually setting castflags to 7\, but I did get failures\, and I did see the same warnings\.

Your changes fixed both the compiler warnings and the test failures\.

Applied your changes as 071db91b12fc08b6b110d18071c50fbe490129ce and added you to AUTHORS in 966178a1f62804c2ffcbd4cb0d2e9a28a118ea1c\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this report\.  You have helped make Perl better\.
&nbsp;    
With the release of Perl 5\.24\.0 on May 9\, 2016\, this and 149 other issues have been resolved\.

Perl 5\.24\.0 may be downloaded via https://metacpan.org/release/RJBS/perl-5.24.0

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
